scripts/irods/paths.py

```diff
@@ -1,139 +1,452 @@
+
+    # In get_irods_directory_impl, we use std::filesystem::path::lexically_relative to produce a
+    # relative path that can be used to traverse from CMAKE_INSTALL_LIBDIR to install_path.
+    # This path usually starts with one or more '..' components.
+    # unlike std::filesystem::path::lexically_relative, pathlib.PurePath.relative_to will throw
+    # ValueError if the calling PurePath instance (self) is not a subpath of the PurePath passed in.
+
+    len_install_dir = len(install_path.parts)
+    for common_part_qty in range(min(len_install_dir, len_irods_dir)):
+        if install_dir.parts[common_part_qty] != irods_dir.parts[common_part_qty]:
+            break
+    else:
+        common_part_qty += 1
+
+    if common_part_qty == 0:
+        # if there is no common base, we can't be smart about it
+        return pathlib.Path(*irods_dir.parts[:-len_irods_dir], install_path)
+
+    uncommon_part_qty = len_irods_dir - common_part_qty
+    return pathlib.Path(*irods_dir.parts[:-uncommon_part_qty], *install_path.parts[common_part_qty:])
+
+_irods_directory_cache = None
+def _irods_directory():
+    global _irods_directory_cache
+    if _irods_directory_cache is None:
+        _irods_directory_cache = _scripts_directory().parent
+    return _irods_directory_cache
 
 def irods_directory():
-    return os.path.join(root_directory(), 'var', 'lib', 'irods')
+    return str(_irods_directory())
+
+_sysconf_directory_cache = None
+def _sysconf_directory():
+    global _sysconf_directory_cache
+    if _sysconf_directory_cache is None:
+        _sysconf_directory_cache = _convert_from_install_dir(_sysconf_install_directory())
+    return _sysconf_directory_cache
+
+def sysconf_directory():
+    return str(_sysconf_directory())
+
+_config_directory_cache = None
+def _config_directory():
+    global _config_directory_cache
+    if _config_directory_cache is None:
+        _config_directory_cache = _sysconf_directory() / 'irods'
+    return _config_directory_cache
 
 def config_directory():
-    return os.path.join(root_directory(), 'etc', 'irods')
+    return str(_config_directory())
+
+_plugins_directory_cache = None
+def _plugins_directory():
+    global _plugins_directory_cache
+    if _plugins_directory_cache is None:
+        _plugins_directory_cache = _convert_from_install_dir(_plugins_install_directory())
+    return _plugins_directory_cache
 
 def plugins_directory():
-    return os.path.join(root_directory(), 'usr', 'lib', 'irods', 'plugins')
+    return str(_plugins_directory())
 
+# TODO: is this what we really want here?
+# Not cached, as irods_user_and_group_entries return value could change
+def _home_directory():
+    return pathlib.Path(irods_user_and_group_entries()[0].pw_dir)
+
+# TODO: is this what we really want here?
+# Not cached, as irods_user_and_group_entries return value could change
 def home_directory():
-    return irods_user_and_group_entries()[0][5]
+    return irods_user_and_group_entries()[0].pw_dir
```

Comment:
`home_directory()` always returns the home directory of the irods service account, rather than the home directory of the user executing the script. Therefore, `default_client_environment_path()` and friends always return the path to the irods service account's files, which may not always be desirable.
I think we may want to consider renaming these methods to convey that the paths returned are specifically for the irods service account.

Reply:
I've changed `# TODO: is this what we really want here?` to be more specific about this